File: ios/VectorialData/VectorialData/Core/Notifications/NotificationsManager.swift
import Foundation
import UIKit
import UserNotifications

/// Owns the push notification lifecycle.
///
/// Flow:
///   1. `requestAuthorization` asks the user and registers with APNs.
///   2. `AppDelegate` forwards `didRegisterForRemoteNotificationsWithDeviceToken`
///      → `didReceiveToken(_:)` which POSTs to `/api/notifications/register-device`.
///   3. On sign out call `unregister` which DELETEs the token.
@MainActor
final class NotificationsManager: NSObject, ObservableObject {
    static let shared = NotificationsManager()

    @Published private(set) var authorizationStatus: UNAuthorizationStatus = .notDetermined
    @Published private(set) var lastRegisteredToken: String?

    /// The APNs token persists across launches so `unregister()` can DELETE
    /// it on sign-out even after a cold start (the in-memory copy is gone by
    /// then). Stored in UserDefaults — it's a device token, not a secret.
    private let tokenDefaultsKey = "apns.deviceToken"

    /// When the user taps a "Nuevo pick" push the payload arrives here. The
    /// Picks tab consumes and clears it via `consumePendingPickTap()`.
    @Published var pendingPickNumber: Int?

    /// Fires when the user taps a Friday-digest push. The weekly digest
    /// screen consumes and clears it.
    @Published var pendingWeeklyDigest: Bool = false

    /// Set when the user taps a curated-news push. HomeView consumes
    /// and clears it after pushing the detail onto its nav stack.
    @Published var pendingNewsId: UUID?

    /// Triggers the paywall sheet in PicksView via URL scheme. Consumed
    /// immediately after presentation; not persisted across launches.
    @Published var pendingShowPaywall: Bool = false

    /// Set when the user taps the scheduled "raise your amount" reminder, so
    /// MainTabView opens the per-buy amount editor.
    @Published var pendingOpenAmount: Bool = false

    private override init() {
        super.init()
        UNUserNotificationCenter.current().delegate = self
        Task { await refreshStatus() }
    }

    func refreshStatus() async {
        let settings = await UNUserNotificationCenter.current().notificationSettings()
        authorizationStatus = settings.authorizationStatus
    }

    /// Ask the user and register for remote notifications. Returns whether
    /// authorization was granted.
    @discardableResult
    func requestAuthorization() async -> Bool {
        do {
            let granted = try await UNUserNotificationCenter.current()
                .requestAuthorization(options: [.alert, .badge, .sound])
            await refreshStatus()
            if granted {
                UIApplication.shared.registerForRemoteNotifications()
            }
            return granted
        } catch {
            return false
        }
    }

    /// Called by `AppDelegate` with the raw APNs token bytes.
    func didReceiveToken(_ data: Data) {
        let token = data.map { String(format: "%02x", $0) }.joined()
        lastRegisteredToken = token
        UserDefaults.standard.set(token, forKey: tokenDefaultsKey)
        Task { await register(token: token) }
    }

    /// Re-attach this device's push token to the signed-in user. Called after
    /// a successful profile load so a token minted while signed out (or under
    /// the previous user) gets bound to the current account.
    func refreshRegistrationIfAuthorized() async {
        guard let token = lastRegisteredToken
            ?? UserDefaults.standard.string(forKey: tokenDefaultsKey)
        else { return }
        lastRegisteredToken = token
        await register(token: token)
    }

    /// Clears in-memory deep-link payloads. Called on sign-out so a tapped
    /// push from the previous session doesn't route the next user.
    func clearPending() {
        pendingPickNumber = nil
        pendingWeeklyDigest = false
        pendingNewsId = nil
        pendingShowPaywall = false
    }

    /// Call on sign-out to stop receiving pushes on this device. Reads the
    /// token from persistence so it works even after a cold launch where the
    /// in-memory copy was never repopulated.
    func unregister() async {
        guard let token = lastRegisteredToken
            ?? UserDefaults.standard.string(forKey: tokenDefaultsKey)
        else { return }
        struct Body: Encodable { let token: String }
        _ = try? await APIClient.shared.delete(
            "/api/notifications/register-device",
            body: Body(token: token),
            as: EmptyResponse.self
        )
        lastRegisteredToken = nil
        UserDefaults.standard.removeObject(forKey: tokenDefaultsKey)
    }

    private func register(token: String) async {
        // Only bind the token to a user once we actually have a session;
        // otherwise the unauthenticated POST 401s and the token is dropped.
        guard AuthManager.shared.state == .signedIn else { return }
        struct Body: Encodable {
            let token: String
            let platform: String
            let appVersion: String?
        }
        let version = Bundle.main
            .object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
        _ = try? await APIClient.shared.post(
            "/api/notifications/register-device",
            body: Body(token: token, platform: "ios", appVersion: version),
            as: EmptyResponse.self
        )
    }
}

extension NotificationsManager: UNUserNotificationCenterDelegate {
    // Show notifications while app is foregrounded.
    nonisolated func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification
    ) async -> UNNotificationPresentationOptions {
        [.banner, .sound, .badge]
    }

    nonisolated func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse
    ) async {
        // Parse everything into Sendable locals here, in the nonisolated
        // context, so the closure hopped to the main actor never captures the
        // non-Sendable `userInfo` ([AnyHashable: Any]).
        let userInfo = response.notification.request.content.userInfo
        let kind = userInfo["kind"] as? String
        let pickNumber = parsePickNumber(userInfo)
        let newsId = (userInfo["news_id"] as? String).flatMap(UUID.init(uuidString:))

        await MainActor.run {
            switch kind {
            case "new_pick":
                if let pickNumber { Self.shared.pendingPickNumber = pickNumber }
            case "weekly_digest":
                Self.shared.pendingWeeklyDigest = true
            case "news":
                if let newsId { Self.shared.pendingNewsId = newsId }
            case "dividend_paid":
                // Re-use the new-pick deep link: route to the pick detail
                // where the new "DIVIDENDOS" section is already visible.
                if let pickNumber { Self.shared.pendingPickNumber = pickNumber }
            case "raise_amount":
                Self.shared.pendingOpenAmount = true
            default:
                break
            }
        }
    }

    /// Reads `pick_number` from a notification payload, tolerating both Int and
    /// String encodings. Pure + nonisolated so it runs in the delegate context.
    private nonisolated func parsePickNumber(_ userInfo: [AnyHashable: Any]) -> Int? {
        if let n = userInfo["pick_number"] as? Int { return n }
        if let s = userInfo["pick_number"] as? String { return Int(s) }
        return nil
    }
}

File: android/app/src/main/java/com/vectorialdata/app/core/notifications/NotificationsManager.kt
package com.vectorialdata.app.core.notifications

import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.os.Build
import androidx.core.app.NotificationManagerCompat
import com.vectorialdata.app.BuildConfig
import com.vectorialdata.app.R
import com.vectorialdata.app.core.auth.AuthManager
import com.vectorialdata.app.core.auth.SecureStore
import com.vectorialdata.app.core.net.ApiClient
import com.vectorialdata.app.core.net.EmptyResponse
import com.google.firebase.messaging.FirebaseMessaging
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.Serializable
import kotlin.coroutines.resume

/**
 * Owns the push notification lifecycle — direct port of the iOS
 * `NotificationsManager`.
 *
 * Flow:
 *   1. Account screen requests POST_NOTIFICATIONS (Android 13+); once granted,
 *      [refreshRegistrationIfEnabled] fetches the FCM token and POSTs it to
 *      `/api/notifications/register-device` (platform "android").
 *   2. [PushMessagingService] forwards token rotations via [didReceiveToken].
 *   3. On sign-out `AuthManager` calls [unregister], which DELETEs the token.
 *
 * Tapping a notification lands in `MainActivity`, which parses the payload
 * into the `pending*` flows below; the tab scaffold and screens consume them
 * (same contract as the iOS `@Published pending*` properties).
 */
object NotificationsManager {
    /** Channel id referenced by the backend sender (`push.ts`). */
    const val CHANNEL_ID = "vd_default"

    /**
     * The FCM token persists so [unregister] can DELETE it on sign-out even
     * after a cold start. Lives in the already-initialized encrypted prefs
     * (iOS keeps it in UserDefaults — it's a device token, not a secret, but
     * SecureStore is the store we already have).
     */
    private const val TOKEN_KEY = "push.fcmToken"

    private lateinit var appContext: Context
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    private val _enabled = MutableStateFlow(false)
    /** Whether notifications are currently allowed for the app. */
    val enabled: StateFlow<Boolean> = _enabled.asStateFlow()

    /** Set when the user taps a "new pick"/"dividend" push; Picks tab consumes it. */
    val pendingPickNumber = MutableStateFlow<Int?>(null)

    /** Set when the user taps the Friday-digest push; Picks tab consumes it. */
    val pendingWeeklyDigest = MutableStateFlow(false)

    /** Set when the user taps a curated-news push; Home consumes it. */
    val pendingNewsId = MutableStateFlow<String?>(null)

    /** Called once from [com.vectorialdata.app.VectorialDataApp]. */
    fun init(context: Context) {
        appContext = context.applicationContext
        createChannel()
        refreshStatus()
    }

    fun refreshStatus() {
        _enabled.value = NotificationManagerCompat.from(appContext).areNotificationsEnabled()
    }

    /** Called by [PushMessagingService] when FCM mints or rotates the token. */
    fun didReceiveToken(token: String) {
        SecureStore.set(token, TOKEN_KEY)
        scope.launch { register(token) }
    }

    /**
     * Re-attach this device's token to the signed-in user. Called after a
     * successful profile load so a token minted while signed out (or under
     * the previous user) gets bound to the current account.
     */
    suspend fun refreshRegistrationIfEnabled() {
        refreshStatus()
        if (!_enabled.value) return
        val token = currentToken() ?: SecureStore.get(TOKEN_KEY) ?: return
        SecureStore.set(token, TOKEN_KEY)
        register(token)
    }

    /**
     * Call on sign-out (while the bearer is still valid) to stop receiving
     * the signed-out user's notifications on this device. Never throws — a
     * failed unregister must not block the sign-out.
     */
    suspend fun unregister() {
        val token = SecureStore.get(TOKEN_KEY) ?: return
        try {
            ApiClient.delete<UnregisterBody, EmptyResponse>(
                "/api/notifications/register-device",
                UnregisterBody(token),
            )
        } catch (_: Exception) {
            // Best effort; dead tokens also get pruned server-side on send.
        }
        SecureStore.delete(TOKEN_KEY)
    }

    /**
     * Clears pending deep-link payloads. Called on sign-out so a tapped push
     * from the previous session doesn't route the next user.
     */
    fun clearPending() {
        pendingPickNumber.value = null
        pendingWeeklyDigest.value = false
        pendingNewsId.value = null
    }

    /**
     * Routes a tapped notification's data payload (all values arrive as
     * strings from FCM) into the pending flows.
     */
    fun handlePushTap(kind: String?, pickNumber: Int?, newsId: String?) {
        when (kind) {
            // dividend_paid reuses the pick deep link, same as iOS: the pick
            // detail already shows the DIVIDENDOS section.
            "new_pick", "dividend_paid" -> pickNumber?.let { pendingPickNumber.value = it }
            "weekly_digest" -> pendingWeeklyDigest.value = true
            "news" -> newsId?.let { pendingNewsId.value = it }
        }
    }

    // ---- Internals ------------------------------------------------------------

    @Serializable
    private data class RegisterBody(val token: String, val platform: String, val appVersion: String?)

    @Serializable
    private data class UnregisterBody(val token: String)

    private suspend fun register(token: String) {
        // Only bind the token to a user once we actually have a session;
        // otherwise the unauthenticated POST 401s and the token is dropped.
        if (AuthManager.state.value != AuthManager.AuthState.SIGNED_IN) return
        try {
            ApiClient.post<RegisterBody, EmptyResponse>(
                "/api/notifications/register-device",
                RegisterBody(token, platform = "android", appVersion = BuildConfig.VERSION_NAME),
            )
        } catch (_: Exception) {
            // Non-fatal: the next profile load retries via refreshRegistrationIfEnabled.
        }
    }

    private suspend fun currentToken(): String? = suspendCancellableCoroutine { cont ->
        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
            cont.resume(if (task.isSuccessful) task.result else null)
        }
    }

    private fun createChannel() {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) return
        val channel = NotificationChannel(
            CHANNEL_ID,
            appContext.getString(R.string.notif_channel_name),
            NotificationManager.IMPORTANCE_HIGH,
        )
        val manager = appContext.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.createNotificationChannel(channel)
    }
}
